File: .skills/openclaw-skills/skills/cheenu1092-oss/tokenmeter/tokenmeter/pricing.py
"""Model pricing data for various AI providers.

Prices are per 1M tokens (input/output) as of Feb 2026.
"""

from dataclasses import dataclass
from typing import Optional


@dataclass
class ModelPricing:
    """Pricing for a specific model."""
    provider: str
    model: str
    input_per_1m: float  # $ per 1M input tokens
    output_per_1m: float  # $ per 1M output tokens
    aliases: tuple[str, ...] = ()  # Alternative names


# Anthropic pricing (as of Feb 2026)
ANTHROPIC_PRICING = [
    ModelPricing("anthropic", "claude-opus-4", 15.00, 75.00, ("opus-4", "opus")),
    ModelPricing("anthropic", "claude-sonnet-4", 3.00, 15.00, ("sonnet-4", "sonnet")),
    ModelPricing("anthropic", "claude-3.5-sonnet", 3.00, 15.00, ("claude-3-5-sonnet", "3.5-sonnet")),
    ModelPricing("anthropic", "claude-3.5-haiku", 0.80, 4.00, ("haiku", "3.5-haiku")),
    ModelPricing("anthropic", "claude-3-opus", 15.00, 75.00, ("claude-3-opus-20240229",)),
]

# OpenAI pricing
OPENAI_PRICING = [
    ModelPricing("openai", "gpt-4o", 2.50, 10.00, ("4o",)),
    ModelPricing("openai", "gpt-4o-mini", 0.15, 0.60, ("4o-mini",)),
    ModelPricing("openai", "gpt-4-turbo", 10.00, 30.00, ("gpt-4-turbo-preview",)),
    ModelPricing("openai", "gpt-4", 30.00, 60.00, ()),
    ModelPricing("openai", "o1", 15.00, 60.00, ("o1-preview",)),
    ModelPricing("openai", "o1-mini", 3.00, 12.00, ()),
    ModelPricing("openai", "o3-mini", 1.10, 4.40, ()),
]

# Google pricing
GOOGLE_PRICING = [
    ModelPricing("google", "gemini-2.0-flash", 0.10, 0.40, ("gemini-flash", "2.0-flash")),
    ModelPricing("google", "gemini-2.0-pro", 1.25, 5.00, ("gemini-pro", "2.0-pro")),
    ModelPricing("google", "gemini-1.5-pro", 1.25, 5.00, ("1.5-pro",)),
    ModelPricing("google", "gemini-1.5-flash", 0.075, 0.30, ("1.5-flash",)),
]

# Azure OpenAI (same as OpenAI for most models)
AZURE_PRICING = [
    ModelPricing("azure", "gpt-4o", 2.50, 10.00, ("azure-gpt-4o",)),
    ModelPricing("azure", "gpt-4o-mini", 0.15, 0.60, ("azure-4o-mini",)),
    ModelPricing("azure", "gpt-4", 30.00, 60.00, ("azure-gpt-4",)),
]

# Combined pricing lookup
ALL_PRICING = ANTHROPIC_PRICING + OPENAI_PRICING + GOOGLE_PRICING + AZURE_PRICING


def get_pricing(provider: str, model: str) -> Optional[ModelPricing]:
    """Look up pricing for a provider/model combo."""
    provider = provider.lower()
    model = model.lower()
    
    for p in ALL_PRICING:
        if p.provider == provider:
            if p.model.lower() == model or model in [a.lower() for a in p.aliases]:
                return p
    
    # Fuzzy match - check if model contains the pricing model name
    for p in ALL_PRICING:
        if p.provider == provider and p.model.lower() in model:
            return p
    
    return None


def calculate_cost(provider: str, model: str, input_tokens: int, output_tokens: int) -> float:
    """Calculate the cost in USD for a given usage."""
    pricing = get_pricing(provider, model)
    if not pricing:
        return 0.0
    
    input_cost = (input_tokens / 1_000_000) * pricing.input_per_1m
    output_cost = (output_tokens / 1_000_000) * pricing.output_per_1m
    return input_cost + output_cost


def list_supported_models() -> list[tuple[str, str]]:
    """Return list of (provider, model) tuples."""
    return [(p.provider, p.model) for p in ALL_PRICING]

File: .skills/openclaw-skills/skills/someone0070/clawmeter/src/pricing.mjs
// Model pricing database — prices per million tokens (USD)
// Last updated: 2026-02-14

export const MODEL_PRICING = {
  // Anthropic
  'claude-opus-4-6':        { input: 15.00, output: 75.00, cacheRead: 1.50, cacheWrite: 18.75 },
  'claude-opus-4':          { input: 15.00, output: 75.00, cacheRead: 1.50, cacheWrite: 18.75 },
  'claude-sonnet-4-5':      { input: 3.00,  output: 15.00, cacheRead: 0.30, cacheWrite: 3.75 },
  'claude-sonnet-4':        { input: 3.00,  output: 15.00, cacheRead: 0.30, cacheWrite: 3.75 },
  'claude-haiku-3-5':       { input: 0.80,  output: 4.00,  cacheRead: 0.08, cacheWrite: 1.00 },
  'claude-3-5-sonnet':      { input: 3.00,  output: 15.00, cacheRead: 0.30, cacheWrite: 3.75 },
  'claude-3-5-haiku':       { input: 0.80,  output: 4.00,  cacheRead: 0.08, cacheWrite: 1.00 },
  'claude-3-opus':          { input: 15.00, output: 75.00, cacheRead: 1.50, cacheWrite: 18.75 },

  // OpenAI
  'gpt-4o':                 { input: 2.50,  output: 10.00 },
  'gpt-4o-mini':            { input: 0.15,  output: 0.60 },
  'gpt-4-turbo':            { input: 10.00, output: 30.00 },
  'gpt-4':                  { input: 30.00, output: 60.00 },
  'o1':                     { input: 15.00, output: 60.00 },
  'o1-mini':                { input: 3.00,  output: 12.00 },
  'o3':                     { input: 10.00, output: 40.00 },
  'o3-mini':                { input: 1.10,  output: 4.40 },
  'o4-mini':                { input: 1.10,  output: 4.40 },

  // Google
  'gemini-2.0-flash':       { input: 0.10,  output: 0.40 },
  'gemini-2.0-pro':         { input: 1.25,  output: 5.00 },
  'gemini-1.5-pro':         { input: 1.25,  output: 5.00 },
  'gemini-1.5-flash':       { input: 0.075, output: 0.30 },

  // DeepSeek
  'deepseek-chat':          { input: 0.27,  output: 1.10 },
  'deepseek-reasoner':      { input: 0.55,  output: 2.19 },
};

// Fuzzy match: try exact, then prefix match, then contains
export function getModelPricing(modelId) {
  if (!modelId) return null;
  const id = modelId.toLowerCase();

  // Exact match
  if (MODEL_PRICING[id]) return MODEL_PRICING[id];

  // Try matching keys
  for (const [key, pricing] of Object.entries(MODEL_PRICING)) {
    if (id.includes(key) || key.includes(id)) return pricing;
  }

  return null;
}

// Calculate cost from usage object (as found in session logs)
export function calculateCost(usage, modelId) {
  // If the log already has cost data, use it
  if (usage.cost && typeof usage.cost.total === 'number') {
    return usage.cost;
  }

  const pricing = getModelPricing(modelId);
  if (!pricing) {
    return { input: 0, output: 0, cacheRead: 0, cacheWrite: 0, total: 0, estimated: true, unknownModel: modelId };
  }

  const inputCost = (usage.input || 0) / 1_000_000 * pricing.input;
  const outputCost = (usage.output || 0) / 1_000_000 * pricing.output;
  const cacheReadCost = (usage.cacheRead || 0) / 1_000_000 * (pricing.cacheRead || 0);
  const cacheWriteCost = (usage.cacheWrite || 0) / 1_000_000 * (pricing.cacheWrite || 0);
  const total = inputCost + outputCost + cacheReadCost + cacheWriteCost;

  return { input: inputCost, output: outputCost, cacheRead: cacheReadCost, cacheWrite: cacheWriteCost, total, estimated: true };
}
